Clarification re repeated addContextListener and addIntentListener calls

kemerava opened the pull request:
Closes #1390 

Based on the discussion on the issue, adding the clarification to the standard on the expected behavior when multiple calls are made to addContextListener and addIntentListener 


@michael-bowen-sc

kriswest reviewed (changes requested):
Good job! Personally, I think that we triggering of all listeners is currently a MUST and should remain so (meaning the SHOULDs in the this PR become MUSTs to explicitly state that).

I also wonder if we need to add a comment about reconnection. If a web app is refreshed it needs to add new handlers as the old ones will be fone - the DA should be tracking the lifecycle of the app and considering them to have been removed if the app reloads/navigates/reconnects. That generally only applies to web apps as most others don't 'reconnect' or 'refresh' in a way that would wipe out their handler functions

bingenito commented:
I see a mix of resolvererror and resolveerror in text and anchor links.

kriswest commented on docs/api/ref/DesktopAgent.md:
```diff
@@ -264,6 +266,8 @@ The [`PrivateChannel`](PrivateChannel) type is provided to support synchronizati
 
 Optional metadata about each intent & context message received, including the app that originated the message, SHOULD be provided by the desktop agent implementation.
 
+ Adding multiple intent listeners on the same type MUST be rejected with the [`ResolverError.IntentListenerConflict`](Errors#resolveerror), unless the previous listener was removed first though [`listener.unsubscribe`](Types#unsubscribe)
```
```suggestion
 Adding multiple intent listeners on the same type MUST be rejected with the [`ResolveError.IntentListenerConflict`](Errors#resolveerror), unless the previous listener was removed first though [`listener.unsubscribe`](Types#unsubscribe)
```

kriswest commented on src/api/Channel.ts:
```diff
@@ -70,6 +70,9 @@ export interface Channel {
    *
    * Optional metadata about each context message received, including the app that originated the message, SHOULD be provided by the desktop agent implementation.
+   * 
+   * Adding multiple context listeners on the same or overlapping types (i.e. named type and null type) SHOULD be allowed, and SHOULD trigger all contextHandlers when a relevant context type is broadcast on the current channel.
```
```suggestion
   * Adding multiple context listeners on the same or overlapping types (i.e. named type and null type) SHOULD be allowed, and SHOULD trigger all context handlers when a relevant context type is broadcast on the current channel.
```

kriswest commented on src/api/DesktopAgent.ts:
```diff
@@ -305,6 +305,8 @@ export interface DesktopAgent {
    *
    * Optional metadata about the raised intent, including the app that originated the message, SHOULD be provided by the desktop agent implementation.
+   * 
+   * Adding multiple intent listeners on the same type MUST be rejected with the `ResolverError.IntentListenerConflict`, unless the previous listener was removed first though `listener.unsubscribe`
```
```suggestion
   * Adding multiple intent listeners on the same type MUST be rejected with the `ResolveError.IntentListenerConflict`, unless the previous listener was removed first though `listener.unsubscribe`
```

kriswest commented:
> I see a mix of resolvererror and resolveerror in text and anchor links.

Good 👀 @bingenito - I missed several of those. Added to suggestions

kemerava commented on docs/api/ref/Channel.md:
```diff
@@ -174,6 +174,8 @@ If, when this function is called, the channel already contains context that woul
 
 Optional metadata about each context message received, including the app that originated the message, SHOULD be provided by the desktop agent implementation.
 
+Adding multiple context listeners on the same or overlapping types (i.e. specific `contextType` and `null` type) SHOULD be allowed, and SHOULD trigger all ContextHandlers when a relevant context type is broadcast on the current channel. 
```
Changed the SHOULD to MUST - @kriswest would you like to wait until the tests are in the conformance to close this comment?

kemerava replied:
I also added the test definitions to the files you mentioned

kemerava replied:
@kriswest let me know if anything else needs adding here or if it can be resolved

kemerava commented on src/api/DesktopAgent.ts:
```diff
@@ -359,6 +361,8 @@ export interface DesktopAgent {
    *
    * Optional metadata about the context message, including the app that originated the message, SHOULD be provided by the desktop agent implementation.
+   * 
+   * Adding multiple context listeners on the same or overlapping types (i.e. named type and null type) SHOULD be allowed, and SHOULD trigger all contextHandlers when a relevant context type is broadcast on the current user channel. Please note, that this behavior differes from `fdc3.addIntentListener`API call; refer to the relevant documentation for more detials. 
```
Changed

kriswest commented:
> @kriswest In terms of the reconnection, I think that there a case which can to be made for Desktop Agent to keep the listeners in memory at least for a short period of time. Not sure as much for web apps, but for others, depending on which method of connection is used, it is possible that the app temporarily lost the connection, and I am not sure if it makes sense to always ask apps to reconnect.
> This is my scenario, for example, if Desktop Agent is determining the app connection based on the heartbeats from the app, and then the app disconnects, then the behavior is slightly undererministic, where Desktop Agent might have already "noticed" that the app has disconnected, and removed the listeners, or it has not, and then the listeners are still there, and then the app unnecessarily sending more calls to DA. Of course, even with keeping listeners in memory we cannot guarantee for sure if that senario would not happen. We can also think about introducing an endpoint which would list all the listenerIds which DA has for the given app, this way the app can verify if it is missing some listeners or if there are some which are no longer needed and can be cleaned up
> Please let me know what you think!

Apologies, for the delay in responding to your question on this. Disconnection is an awkward topic for FDC3 as what it means could vary greatly between implementations. In the classic container implementations for web apps, its doesn't really happen other than via the lifecycle of the application. That's because the Desktop Agent tends to be provided by a local service of some sort and is therefore not affected by the network connection... If it goes down, the Desktop Agent API is still available (the connection between the app and Desktop Agent is unaffected), listeners remain valid etc.. 

I suspect that this will be the same for most FDC3 for the Web implementations... if the Desktop Agent was already loaded and connected to, then the network connection going down doesn't affect its ability to provide interoperability between windows! Of course, there will be exceptions: Desktop Agent implementations which rely on a remote web service and situations where the Desktop Agent window is refreshed and reloads. I suspect these cases need thinking about. In both cases there is potentially a need for some sort of event to pass back for either the apps or the `getAgent` code to handle. A new connection to the DA needs to be set up (as we'll lose the Message Ports in the later case).

I think we'd be well advised to trigger the connection flow afresh - the DA might come back with a different config for apps for example. The alternative is that the Desktop Agent. Whether the getAgent code retains listener or requires new ones to be set-up at that point is something we'll need to consider. `getAgent()` will certainly need to re-register the listeners with the DA, even if the app doesn't have to re-add them.

When it comes to native apps and disconnection, we will usually be talking about a local service that provides access to the DA going down or being restarted. In those cases, new connections (and polling for connection) will likely be needed. Whether listeners need to be readded or not feels like something we should discuss as it should probably be implemented consistently (and hence handled in a standardized way).

At the very least we'll need to think about adding an (`addEventListener`) event for disconnect and reconnect and then discuss how that should be handled. If what I've said here makes sense, would you be up for raising an issue for the Standards Working Group to discuss? It  might help to enumerate all the different types of disconnection that could occur and figure out which need solutions (I think the DA going away, at least temporarily, is the primary one to address).

kriswest commented:
P.S. I think this PR is ready to go - the only thing holding it up is the fact that we closed the 2.2 scope and this does introduce a new error that MUST be thrown. However, it's not a big change and I would be happy to consider a motion to put it into 2.2 (which is currently blocked by completing the `getAgent` implementation work) if participants wanted to get it implemented and there were no objections or existing examples of alternative ways of handling intent listener conflicts that another participant wanted to to preserve.

kemerava commented:
> P.S. I think this PR is ready to go - the only thing holding it up is the fact that we closed the 2.2 scope and this does introduce a new error that MUST be thrown. However, it's not a big change and I would be happy to consider a motion to put it into 2.2 (which is currently blocked by completing the `getAgent` implementation work) if participants wanted to get it implemented and there were no objections or existing examples of alternative ways of handling intent listener conflicts that another participant wanted to to preserve.

@kriswest has there been consensus on this one? In my view, it is a pretty small change, but if we want to hold it off until 2.3 that is fine as well

kriswest commented:
We were set to hold off for 2.3 for purely procedural reasons (we voted to close the scope). However, I agree this is small and fixes an existing problem and could go in for 2.2. This PR could be called a breaking change as it introduces a new error that could be returned on existing calls to `raiseIntent` and `raiseIntentForContext`. However, you would previously have been faced with unstandardized behaviour if you tried to add a listener for the same intent multiple times and DAs might have returned a non-standard error, overwritten the handler or something else. Hence, I think bringing this in is a better situation.

Regardless it will need approval by the SWG which we can do at the next meeting.

kemerava commented on website/versioned_docs/version-2.0/api/conformance/App-Channel-Tests.md:
```diff
@@ -59,3 +59,18 @@ hide_title: true
 - `ACContextHistoryLast`: In step 5. **A** retrieves the _untyped_ current context of the channel via `const currentContext = await testChannel.getCurrentContext()`. Ensure that A receives only the very last broadcast context item _of any type_.
+
+
+## Multipe listeners On The Same Or Overlapping Ccontext types 
```
Done, not sure why they ended up there

kemerava commented on website/versioned_docs/version-2.1/api/conformance/User-Channel-Tests.md:
```diff
@@ -55,3 +55,17 @@ _NB:  User Channels were called System Channels in FDC3 1.2.  The new terminolog
  - `UCFilteredUsageNoJoin`: Perform the above test, but skip step 2 so that **A** does NOT join a channel. Confirm that the _current channel_ for **A** is NOT set before continuing with the rest of the test.  **A** should receive nothing.
+
+
+## Broadcast With Multiple Listeners On The Same or Overlapping Types
```
Done, thanks!

kemerava commented:
> Conformance test additions need moving into the next version of the docs. Also I think we're missing a test that triggers `IntentListenerConflict`.

@kriswest thank you for the review! Added the tests and fixed the location of the previous ones. Could you please re-review? Thanks!

kriswest reviewed (approved):
LGTM - I wondered whether anything needed to be added to the compliance table at: https://fdc3.finos.org/docs/api/spec#desktop-agent-api-standard-compliance
for the new MUSTs added in this PR. However, on reviewing that list I note that existing points:

> - Include implementations of the following [Desktop Agent](https://fdc3.finos.org/docs/api/ref/DesktopAgent) API functions, as defined in this Standard

...

> - Return (JavaScript or platform appropriate) Error Objects with messages from the [ChannelError](https://fdc3.finos.org/docs/api/ref/Errors#channelerror), [OpenError](https://fdc3.finos.org/docs/api/ref/Errors#openerror), [ResolveError](https://fdc3.finos.org/docs/api/ref/Errors#resolveerror) and [ResultError](https://fdc3.finos.org/docs/api/ref/Errors#resulterror) enumerations as appropriate.

Already import the requirement from the reference material and that is how it is dealt with for other similar requirements. hence, I think good to merge after review by @finos/fdc3-editors (@novavi or @hughtroeger could you take a look please)

novavi commented on website/docs/api/conformance/User-Channel-Tests.md:
```diff
@@ -70,3 +70,17 @@ As the method of setting the user channel is user interactive, it is either diff
+
+| App | Step               | Details                                                                                                     |
+|-----|--------------------|-------------------------------------------------------------------------------------------------------------|
+| A   | 1.addContextListeners | A sets up two Context Listeners.  One _untyped_ and one for `fdc3.contact` by calling:  `addContextListener (null, handler1)` <br/> `addContextListener ("fdc3.contact", handler2)` <br/>![1.2](https://img.shields.io/badge/FDC3-1.2-green) A `Listener` object is returned for each.  <br />![2.0](https://img.shields.io/badge/FDC3-2.0-blue) A promise resolving a `Listene ...
```
Remove spaces between method names and open braces:  `addContextListener (null, handler1)` and `addContextListener ("fdc3.contact", handler2)`

novavi commented on website/versioned_docs/version-2.2/api/conformance/App-Channel-Tests.md:
```diff
@@ -36,7 +36,7 @@ hide_title: true
 |-----|-------------------------|-----------------------------------------------------------------|
 | A   | 1. Retrieve `Channel`   | Retrieve a `Channel` object representing an 'App' channel called `test-channel` using: <br />`const testChannel = await fdc3.getOrCreateChannel("test-channel")` |
-| A   | 2. Add Context Listener | Add an _typed_ context listener for `fdc3.instrument`, using: <br />`await testChannel.addContextListener("fdc3.instrument",handler)`|
+| A   | 2. Add Context Listener | Add a _typed_ context listener for `fdc3.instrument`, using: <br />`await testChannel.addContextListener("fdc3.instrument",handler)`|
```
Missing space between comma and argument: `("fdc3.instrument",handler)`

novavi reviewed (approved):
Minor comments

kriswest commented on CHANGELOG.md:
```diff
@@ -41,6 +43,18 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 * Added .NET docs for Events to API reference. ([#1441](https://github.com/finos/FDC3/pull/1441))
 * Setup package publishing for mono-repo packages. ([#1520](https://github.com/finos/FDC3/pull/1520))
 * Implementation PR for FDC3 for the Web ([#896](https://github.com/finos/FDC3/pull/896))
+  - resolves ([#1209](https://github.com/finos/FDC3/issues/1209))
+  - resolves ([#1297](https://github.com/finos/FDC3/issues/1297)) 
+  - resolves ([#1429](https://github.com/finos/FDC3/issues/1429)) 
+  - resolves ([#1430](https://github.com/finos/FDC3/issues/1430))
+  - resolves ([#1431](https://github.com/finos/FDC3/issues/1431))
+  - resolves ([#1432](https://github.com/finos/FDC3/issues/1432)) 
+  - resolves ([#1433](https://github.com/finos/FDC3/issues/1433))
+  - resolves ([#1468](https://github.com/finos/FDC3/issues/1468))
+  - resolves ([#810](https://github.com/finos/FDC3/issues/810))
+  - resolves ([#832](https://github.com/finos/FDC3/issues/832))
+  - resolves ([#1487](https://github.com/finos/FDC3/issues/1487))
+  - resolves ([#1488](https://github.com/finos/FDC3/issues/1488))
```
I think it'd take a rebase to remove this from the diff. These went into main after this PR was started and then leak into the diff after being merged (as it still looks like a diff from the base of the branch). We can ignore

kriswest commented on packages/fdc3-standard/src/api/DesktopAgent.ts:
```diff
@@ -305,6 +305,8 @@ export interface DesktopAgent {
    *
    * Optional metadata about the raised intent, including the app that originated the message, SHOULD be provided by the desktop agent implementation.
+   * 
+   * Adding multiple intent listeners on the same type MUST be rejected with the `ResolveError.IntentListenerConflict`, unless the previous listener was removed first though `listener.unsubscribe`
```
```suggestion
   * Adding multiple intent listeners on the same type MUST be rejected with the `ResolveError.IntentListenerConflict`, unless the previous listener was removed first though `listener.unsubscribe()`
```

kriswest commented on packages/fdc3-standard/src/api/DesktopAgent.ts:
```diff
@@ -359,6 +361,8 @@ export interface DesktopAgent {
    *
    * Optional metadata about the context message, including the app that originated the message, SHOULD be provided by the desktop agent implementation.
+   * 
+   * Adding multiple context listeners on the same or overlapping types (i.e. named type and null type) MUST be allowed, and MUST trigger all contextHandlers when a relevant context type is broadcast on the current user channel. Please note, that this behavior differes from `fdc3.addIntentListener`API call; refer to the relevant documentation for more detials. 
```
```suggestion
   * Adding multiple context listeners on the same or overlapping types (i.e. named type and null type) MUST be allowed, and MUST trigger all contextHandlers when a relevant context type is broadcast on the current user channel. Please note, that this behavior differes from `fdc3.addIntentListener` API call; refer to the relevant documentation for more detials. 
```

kriswest commented on packages/fdc3-standard/src/api/Errors.ts:
```diff
@@ -84,6 +84,9 @@ export enum ResolveError {
   MalformedContext = 'MalformedContext',
 
+  /** Returned if `fdc3.addIntentListener` is called for a specified intent that the application has already added a listener for and has not subsequently removed it. */
+  IntentListenerConflict = "IntentListenerConflict",
```
```suggestion
  IntentListenerConflict = 'IntentListenerConflict',
```

kriswest commented on website/docs/api/conformance/App-Channel-Tests.md:
```diff
@@ -60,3 +60,18 @@ hide_title: true
+
+## Multiple listeners On The Same Or Overlapping Context types  
+
+| App | Step               | Details                                                                    |
+|-----|--------------------|----------------------------------------------------------------------------|
+| A   | 1.Retrieve `Channel`    |Retrieve a `Channel` object representing an 'App' channel called `test-channel` using: <br/>`const testChannel = await fdc3.getOrCreateChannel("test-channel")` |
+| A   | 2.Add Context Listener |Add an _untyped_ context listener to the channel, using: <br/> ![2.0](https://img.shields.io/badge/FDC3-2.0-blue) `await testChannel.addContextListener(null, handler1)` <br/>![1.2](https://img.shields.io/badge/FDC3-1.2-green) `testChannel.addContextListener(null, handler1)` |
+| A   | 3.Add Context Listener |Add a _typed_ context listener for `fdc3.instrument` with a different handler, using: <br/> ![2.0](https://img.shields.io/badge/FDC3-2.0-blue) `await testChannel.addContextListener("fdc3.instrument", handler2)` <br/>![1.2](https://img.shields.io/badge/FDC3-1.2-green) `testChannel.addContextListener("fdc3.instrument", handler2)`|
+| B   | 4.Retrieve `Channel`     | Retrieve a `Channel` object representing the same 'App' channel A did (`test-channel`)|
+| B   | 5.Broadcast          | Broadcast an `fdc3.instrument` Context to the channel with: <br/>`testChannel.broadcast(<fdc3.instrument context>)`|
+| A   | 6.Receive Context    | The handlers added in step 2 and 3 will receive the instrument context. Ensure that the instrument received by A is identical to that sent by B.  |
```
```suggestion
| App | Step               | Details                                                                    |
|-----|--------------------|----------------------------------------------------------------------------|
| A   | 1. Retrieve `Channel`    |Retrieve a `Channel` object representing an 'App' channel called `test-channel` using: <br/>`const testChannel = await fdc3.getOrCreateChannel("test-channel")` |
| A   | 2. Add Context Listener |Add an _untyped_ context listener to the channel, using: <br/> ![2.0](https://img.shields.io/badge/FDC3-2.0-blue) `await testChannel.addContextListener(null, handler1)` <br/>![1.2](https://img.shields.io/badge/FDC3-1.2-green) `testChannel.addContextListener(null, handler1)` |
| A   | 3. Add Context Listener |Add a _typed_ context listener for `fdc3.instrument` with a different handler, using: <br/> ![2.0](https://img.shields.io/badge/FDC3-2.0-blue) `await testChannel.addContextListener("fdc3.instrument", handler2)` <br/>![1.2](https://img.shields.io/badge/FDC3-1.2-green) `testChannel.addContextListener("fdc3.instrument", handler2)`|
| B   | 4. Retrieve `Channel`     | Retrieve a `Channel` object representing the same 'App' channel A did (`test-channel`)|
| B   | 5. Broadcast          | Broadcast an `fdc3.instrument` Context to the channel with: <br/>`testChannel.broadcast(<fdc3.instrument context>)`|
| A   | 6. Receive Context    | The handlers added in step 2 and 3 will receive the instrument context. Ensure that the instrument received by A is identical to that sent by B.  |
```

kriswest commented on website/docs/api/conformance/User-Channel-Tests.md:
```diff
@@ -70,3 +70,17 @@ As the method of setting the user channel is user interactive, it is either diff
+| App | Step               | Details                                                                                                     |
+|-----|--------------------|-------------------------------------------------------------------------------------------------------------|
+| A   | 1.addContextListeners | A sets up two Context Listeners.  One _untyped_ and one for `fdc3.contact` by calling:  `addContextListener (null, handler1)` <br/> `addContextListener ("fdc3.contact", handler2)` <br/>![1.2](https://img.shields.io/badge/FDC3-1.2-green) A `Listener` object is returned for each.  <br />![2.0](https://img.shields.io/badge/FDC3-2.0-blue) A promise resolving a `Listene ...
+| A   | 2.joinUserChannel     |A joins the first available user channel using: <br/>![1.2](https://img.shields.io/badge/FDC3-1.2-green) `getSystemChannels()` Check channels are returned. <br/>![2.0](https://img.shields.io/badge/FDC3-2.0-blue) `getUserChannels()` Check **user** channels are returned.<br/>Call `fdc3.joinChannel()` on the first non-global channel.|
+| B   | 3.joinUserChannel     |B joins the same channel as A, via the same process in 2. |
+| B   | 4.Broadcast          |`fdc3.broadcast(<contact context>)` . |
+| A   | 5.Receive Context    | A's `fdc3.contact` object matches the one broadcast  by B, both handlers from step 1 are triggered, and broadcast arrives on the correct listener.   |
```
```suggestion
| A   | 2. joinUserChannel     |A joins the first available user channel using: <br/>![1.2](https://img.shields.io/badge/FDC3-1.2-green) `getSystemChannels()` Check channels are returned. <br/>![2.0](https://img.shields.io/badge/FDC3-2.0-blue) `getUserChannels()` Check **user** channels are returned.<br/>Call `fdc3.joinChannel()` on the first non-global channel.|
| B   | 3. joinUserChannel     |B joins the same channel as A, via the same process in 2. |
| B   | 4. Broadcast          |`fdc3.broadcast(<contact context>)` . |
| A   | 5. Receive Context    | A's `fdc3.contact` object matches the one broadcast  by B, both handlers from step 1 are triggered, and broadcast arrives on the correct listener.   |
```

kriswest commented:
@hughtroeger and @novavi I've applied your suggestions, could one of you sign off again for @finos/fdc3-editors

kriswest commented:
@finos/fdc3-editors (@hughtroeger and @novavi) I think this is just waiting for your sign-off in order to merge.

kriswest reviewed (approved):
LGTM

julianna-ciq reviewed (approved):
lgtm
